components/file_system/src/rate_limiter.rs

```diff
@@ -1,43 +1,363 @@
+    pub fn new(bytes_per_sec: usize, refill_period: Duration) -> PerTypeIORateLimiter {
+        PerTypeIORateLimiter {
+            bytes_per_refill: AtomicUsize::new(calculate_bytes_per_refill(
+                bytes_per_sec,
+                refill_period,
+            )),
+            consumed: AtomicUsize::new(0),
+            refill_period,
+            last_refill_time: Mutex::new(Instant::now()),
+            condv: Condvar::new(),
+            calibrator: None,
+        }
+    }
+
+    pub fn set_bytes_per_sec(&self, bytes_per_sec: usize) {
+        self.bytes_per_refill.store(
+            calculate_bytes_per_refill(bytes_per_sec, self.refill_period),
+            Ordering::Relaxed,
+        );
+    }
+
+    #[allow(dead_code)]
+    pub fn set_calibrator(&mut self, calibrator: BytesCalibrator) {
+        self.calibrator = Some(calibrator);
+    }
+
+    #[inline]
+    fn request_fast(&self, bytes_per_refill: usize, bytes: usize) -> Option<usize> {
+        if self.consumed.load(Ordering::Relaxed) < bytes_per_refill {
+            // Consumed bytes are allowed to be larger than the actual bytes
+            // through when quotas are drained.
+            let before = self.consumed.fetch_add(bytes, Ordering::Relaxed);
+            if before < bytes_per_refill {
+                return Some(std::cmp::min(bytes_per_refill - before, bytes));
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn refill_and_request(&self, bytes: usize) -> usize {
+        let token = std::cmp::min(self.bytes_per_refill.load(Ordering::Relaxed), bytes);
+        self.consumed.store(token, Ordering::Relaxed);
+        if let Some(calibrator) = &self.calibrator {
+            calibrator.reset();
+        }
+        self.condv.notify_all();
+        token
+    }
+
+    pub fn request(&self, bytes: usize, priority: IOPriority) -> usize {
+        let cached_bytes_per_refill = self.bytes_per_refill.load(Ordering::Relaxed);
+        if cached_bytes_per_refill == 0 {
+            return bytes;
+        }
+        if priority == IOPriority::High {
+            self.consumed.fetch_add(bytes, Ordering::Relaxed);
+            // Unlimited requestor don't do refills themselves, therefore the first
+            // limited IO in long period will take penalty for a mandotory refill.
+            return bytes;
+        }
+        loop {
+            if let Some(bytes) = self.request_fast(cached_bytes_per_refill, bytes) {
+                break bytes;
+            }
+            let mut last_refill_time = self.last_refill_time.lock().unwrap();
+            // double check if bytes have been refilled by others
+            if self.consumed.load(Ordering::Relaxed) < cached_bytes_per_refill {
+                continue;
+            }
+            if let Some(calibrator) = &self.calibrator {
+                let calibrated = calibrator.calibrate();
+                self.consumed.store(calibrated, Ordering::Relaxed);
```

Comment:
not quite understand the logic about `calibrator`

Reply:
For read limiter, it calculates bytes consumption based on buffered read initially. When the quotas are drained, it will call calibrator (if any) to calibrate the consumption estimation against some ground truths. This design is to reduce calibrating frequency.